### PR TITLE
Feature/dapp add service worker unit tests

### LINKED ISSUE #2836: dApp - add unit tests for service worker
See [this comment](https://github.com/raiden-network/light-client/issues/2818#issuecomment-868542268) of #2818 at `Unit tests`.

### PR DESCRIPTION
**Pull request by weilbith**
Fixes #2836

**Short description**
Note that these "unit" tests are partially integration tests, as they also test the whole service worker with all modules and also the Workbox dependency. This is on purpose, as we don't have any integration tests, and this way we can be more sure everything works as intended and detect when the dependency usage breaks with updates.

The main suite for the service worker is quite long. I tried to shim the reduce the code per test cases to a minimum but still make them fully express what they are about and how they work. The mass comes from the sheer amount of various cases the service worker must fulfill. If you think this is too big, we could consider to split it.

Many of the added lines of code are part of the big mocking effort for the service worker environment it runs withing a browser. I suppose it is not necessary to invest too much time on this part of the review. But if you are willed to, I would be very happy. The mocks just implement the minimal required parts of the API following mostly their official specification with some shortcuts here and there to keep it simple enough.

As always, and especially for bigger PRs, I recommend to review commit by commit. It protects from getting overwhelmed. Plus it makes all displayed changes belong together with the respective message.

**Definition of Done**

- [x] Steps to manually test the change have been documented
- [x] Acceptance criteria are met
- [ ] Change has been manually tested by the reviewer (dApp) 

**Steps to manually test the change (dApp)**

1. Just run the unit tests...
2. Maybe try that the service worker still works in practice as there were a few code changes to it.

### CODE REVIEW

**Comment by weilbith** [comment]
No CHANGELOG entry necessary for this PR.

**Comment by weilbith** [comment]
How in hell could the coverage drop? This PR is >1.000 loc for new tests...

**Review by andrevmatos — APPROVED** [review]
Some nitpicks and suggestions. Overall, good job here, quite an extensive test suite for an important feature. Thank you for this.

File: raiden-dapp/tests/unit/service-worker/worker/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+      const caches = new MockedCacheStorage();
+      const context = mockEnvironmentForServiceWorker({ caches });
+
+      await sendMessageToServiceWorker(context, 'update');
+
+      expect(caches.delete).toHaveBeenCalledTimes(1);
+      expect(caches.delete).toHaveBeenLastCalledWith('workbox-precache-v2');
+    });
+
+    // TOOD: Deletes database entries
```
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/tests/unit/service-worker/worker/index.spec.ts:203]: I'll add the actual test for it. 🙈

File: raiden-dapp/tests/unit/utils/mocks/service-worker/environment.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  MockedResponse,
+} from './';
+
+type CacheStorageEnvironment = {
+  cache: MockedCache;
+  caches: MockedCacheStorage;
+};
+
+/**
+ * Please make sure to call `jest.restoreAllMocks()` to "tear down" the
```
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/tests/unit/utils/mocks/service-worker/environment.ts:23]: Okay, lesson learned. I'll see how I can use that in future. Thank you very much for explaining that! 🙏🏾

File: raiden-dapp/src/service-worker/worker/database.js [hunk 3]
```diff
@@ -1,27 +1,53 @@
...
+    request.addEventListener('upgradeneeded', async () => {
+      const database = request.result;
+      const objectStore = database.createObjectStore(OBJECT_STORE_NAME);
+      await resolveTransaction(objectStore.transaction);
+      resolve(database);
+    });
   });
 }
 
+async function getPreacheObjectStore(mode = 'readonly') {
```
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/src/service-worker/worker/database.js]: ✔️

File: raiden-dapp/tests/unit/service-worker/worker/clients.spec.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    jest.resetAllMocks();
+  });
+
+  describe('isAnyClientAvailable', () => {
+    test('is false if no client is connected at all', async () => {
+      const clients = new MockedClients([]);
+
+      const available = await isAnyClientAvailable.call({ clients });
+
+      expect(available).toBeFalsy();
```
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/tests/unit/service-worker/worker/clients.spec.ts]: ✔️

File: raiden-dapp/tests/unit/service-worker/worker/clients.spec.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      const clients = new MockedClients([clientOne, clientTwo]);
+
+      const available = await isAnyClientAvailable.call({ clients });
+
+      expect(available).toBeTruthy();
+    });
+  });
+
+  describe('sendMessageToClients()', () => {
+    test('sends message to a to a single connected client', async () => {
```
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/tests/unit/service-worker/worker/clients.spec.ts]: ✔️

File: raiden-dapp/tests/unit/service-worker/worker/clients.spec.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      const client = new MockedClient('client-id');
+      const clients = new MockedClients([client]);
+
+      await sendMessageToClients.call({ clients }, 'message', { data: 1 });
+
+      expect(client.postMessage).toHaveBeenCalledTimes(1);
+      expect(client.postMessage).toHaveBeenLastCalledWith('message', { data: 1 });
+    });
+
+    test('sends message to a to multiple connected clients', async () => {
```
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/tests/unit/service-worker/worker/clients.spec.ts]: ✔️

File: raiden-dapp/tests/unit/service-worker/worker/index.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-dapp/tests/unit/service-worker/worker/index.spec.ts]: ✔️

**Comment by weilbith** [comment]
CI looks good. I'll squash the fixup commits now and then merge.

**Pull request merged**